File: WebAssemblyBrowserApp/WebAssembly_WebGL/Engine/Input.cs
using System.Runtime.InteropServices.JavaScript;

namespace WebAssemblyBrowserApp.Engine;

/// <summary>
/// 键盘 / 鼠标 / 触摸输入。
/// 键名使用 KeyboardEvent.code（如 "ArrowLeft"、"KeyA"、"Space"）。
/// </summary>
public static partial class Input
{
    public const string ArrowLeft = "ArrowLeft";
    public const string ArrowRight = "ArrowRight";
    public const string ArrowUp = "ArrowUp";
    public const string ArrowDown = "ArrowDown";
    public const string Space = "Space";
    public const string Enter = "Enter";
    public const string KeyA = "KeyA";
    public const string KeyD = "KeyD";
    public const string KeyW = "KeyW";
    public const string KeyS = "KeyS";
    public const string KeyR = "KeyR";
    public const string Escape = "Escape";
    public const string Digit1 = "Digit1";
    public const string Digit2 = "Digit2";

    [JSImport("input.init", "main.js")]
    public static partial void Init();

    /// <summary>按键当前是否按住。</summary>
    [JSImport("input.isKeyDown", "main.js")]
    public static partial bool IsKeyDown(string code);

    /// <summary>按键是否在本帧内刚按下（每帧末自动清除）。</summary>
    [JSImport("input.isKeyPressed", "main.js")]
    public static partial bool IsKeyPressed(string code);

    [JSImport("input.mouseX", "main.js")]
    public static partial double MouseX();

    [JSImport("input.mouseY", "main.js")]
    public static partial double MouseY();

    [JSImport("input.isMouseDown", "main.js")]
    public static partial bool IsMouseDown();

    [JSImport("input.isMousePressed", "main.js")]
    public static partial bool IsMousePressed();

    /// <summary>清除本帧的“刚按下”状态，由引擎在每帧末调用。</summary>
    [JSImport("input.endFrame", "main.js")]
    public static partial void EndFrame();
}

File: WebAssemblyBrowserApp/WebAssembly_WebGPU/wwwroot/jsengine/input/input.js
// =====================================================================
// 输入层：键盘 / 鼠标 / 触摸。
// 坐标换算到逻辑分辨率（逻辑像素），与渲染层坐标系一致。
// =====================================================================

import { audio } from '../audio/audio.js'

let _keys = {}
let _pressed = {}
const _mouse = { x: 0, y: 0, down: false, pressed: false }
let _canvasSelector = '#game'

export const input = {
    init() {
        window.addEventListener('keydown', (e) => {
            if (!e.repeat) { _keys[e.code] = true; _pressed[e.code] = true }
            if (['ArrowUp', 'ArrowDown', 'ArrowLeft', 'ArrowRight', 'Space'].includes(e.code)) e.preventDefault()
            audio.ensure()
        })
        window.addEventListener('keyup', (e) => { _keys[e.code] = false })
        window.addEventListener('mousemove', updateMouse)
        window.addEventListener('mousedown', (e) => { _mouse.down = true; _mouse.pressed = true; audio.ensure() })
        window.addEventListener('mouseup', () => { _mouse.down = false })
        window.addEventListener('touchstart', (e) => {
            updateMouse(e.touches[0])
            _mouse.down = true; _mouse.pressed = true; audio.ensure()
        }, { passive: true })
        window.addEventListener('touchmove', (e) => { updateMouse(e.touches[0]) }, { passive: true })
        window.addEventListener('touchend', () => { _mouse.down = false })
        window.addEventListener('blur', () => { _keys = {}; _pressed = {}; _mouse.down = false })
    },

    isKeyDown(code) { return !!_keys[code] },
    isKeyPressed(code) { return !!_pressed[code] },
    mouseX() { return _mouse.x },
    mouseY() { return _mouse.y },
    isMouseDown() { return _mouse.down },
    isMousePressed() { return _mouse.pressed },
    endFrame() { _pressed = {}; _mouse.pressed = false },
}

function updateMouse(e) {
    const cv = document.querySelector(_canvasSelector)
    if (!cv) return
    const rect = cv.getBoundingClientRect()
    // 逻辑分辨率（渲染层挂到 canvas dataset；canvas 物理像素 = 逻辑 × CSS 缩放）
    // 鼠标 CSS 像素 → 逻辑像素：× 逻辑分辨率 / CSS 显示尺寸
    const vw = parseInt(cv.dataset.vw || '960', 10)
    const vh = parseInt(cv.dataset.vh || '540', 10)
    _mouse.x = (e.clientX - rect.left) * (vw / rect.width)
    _mouse.y = (e.clientY - rect.top) * (vh / rect.height)
}
